spec.md

```diff
@@ -244,6 +244,58 @@ A common use of an externally defined module is to represent a Verilog module
 that will be written separately and provided together with FIRRTL-generated
 Verilog to downstream tools.
 
+## Functions
+
+A function has a name, list of arguments, and a list of statements representing
+a data-flow graph of stateless circuit components.  Each argument is specified 
+by its direction, which may be input or output, a name, and the data type of the 
+argument.
+
+Functions participate in width-inference (see [@sec:width-inference]) at 
+invocation sites in the same way modules to at instantiation sites.
+
+Functions cannot contain memories, registers, or instances.  This makes them
+"stateless"
+
+Functions can optional specify that they are "pure" functions.  These do not
+have side-effecting operations (such as printf).
+
+
+``` firrtl
+function MyPureFunction pure :
+  input foo : UInt
+  output bar : UInt
+  bar <= foo
+
+function MyFunction :
+  input clk: Clock
+  input halt: UInt<1>
+  stop(clk, halt, 42) : optional_name
+```
+
+## Intrinsics
+
+An intrinsic is a function. Intrinsics have all the components of a function
+except  a list of statements.  An intrinsic is a firrtl implementation provided 
```

Comment:
what is "a firrtl implementation". You mean "a firrtl compiler implementation"? Is this a term that is defined already in this document?